subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/InstantExecutionReport.kt

```diff
@@ -110,97 +111,90 @@ class InstantExecutionReport(
 
         val fatalError = runWithExceptionHandling(block)
 
-        if (problems.isEmpty()) {
-            require(fatalError == null)
-            return null
+        return synchronized(problems) {
+            when {
+                problems.isEmpty() -> {
+                    require(fatalError == null)
+                    null
+                }
+                fatalError != null -> {
+                    require(fatalError is InstantExecutionException)
+                    fatalError
+                }
+                else -> {
+                    instantExecutionExceptionForErrors()
+                        ?: instantExecutionExceptionForProblems()
+                }
+            }
         }
-
-        return fatalError?.withSuppressed(errors())
-            ?: instantExecutionExceptionForErrors()
-            ?: instantExecutionExceptionForProblems()
     }
 
     private
     fun runWithExceptionHandling(block: () -> Unit): Throwable? {
         try {
             block()
         } catch (e: Throwable) {
-            when (e.cause ?: e) {
-                is TooManyInstantExecutionProblemsException -> return e
-                is StackOverflowError -> add(e)
-                is Error -> throw e
-                else -> add(e)
+            when (val cause = e.maybeUnwrapInvocationTargetException()) {
+                is InstantExecutionException -> return cause
+                is StackOverflowError -> add(cause)
+                is Error -> throw cause
+                else -> add(cause)
             }
         }
         return null
     }
 
     private
-    fun add(e: Throwable) {
+    fun add(e: Throwable) = synchronized(problems) {
         problems.add(
             unknownPropertyError(e.message ?: e.javaClass.name, e)
         )
     }
 
     private
     fun instantExecutionExceptionForErrors(): Throwable? =
-        errors()
-            .takeIf { it.isNotEmpty() }
-            ?.let { errors -> InstantExecutionErrorsException().withSuppressed(errors) }
+        if (errors().isNotEmpty()) InstantExecutionErrorsException(problems, htmlReportFile)
+        else null
 
     private
     fun instantExecutionExceptionForProblems(): Throwable? =
-        if (startParameter.failOnProblems) InstantExecutionProblemsException()
+        if (startParameter.failOnProblems) InstantExecutionProblemsException(problems, htmlReportFile)
         else null
 
     private
-    fun Throwable.withSuppressed(errors: List<PropertyProblem.Error>) = apply {
-        errors.forEach {
-            addSuppressed(it.exception)
+    fun errors() =
+        problems.filterIsInstance<PropertyProblem.Error>()
+
+    private
+    val outputDirectory: File by lazy {
+        startParameter.rootDirectory.resolve(
```

Comment:
this was already here, but it should be using `BuildLayout` (the result of determining the root directory) rather than `StartParameter` (an input into the calculation that determines the root directory).

Reply:
It actually uses `buildLayout.rootDirectory`. Here `startParameter` is a `InstantExecutionStartParameter` not `StartParameter`.